distributed/scheduler.py

```diff
@@ -5472,144 +5490,340 @@ async def _delete_worker_data(self, worker_address, keys):
+        the number of whitelisted keys, or when most keys are replicated or cannot be
+        moved for some other reason.
+
+        Returns list of tuples to feed into _rebalance_move_data:
+
+            - sender worker
+            - recipient worker
+            - task to be transferred
+        """
+        ts: TaskState
+        ws: WorkerState
 
-                for k, v in workers_by_task.items():
-                    for vv in v:
-                        tasks_by_worker[vv].add(k)
+        # Smallest-first heapsof workers that need to send/receive data, with how many
+        # bytes each needs to send/receive. Bytes are negative, so that the workers
+        # farthest from the cluster mean are at the top of the heaps.
+        # Heap elements:
+        # - maximum number of bytes to send or receive
+        # - number of bytes after which the worker should not be considered anymore
+        # - arbitrary unique number, there just to to make sure that WorkerState objects
+        #   are never used for sorting in the unlikely event that two processes have
+        #   exactly the same number of bytes allocated.
+        # - WorkerState
+        # - iterator of all tasks in memory on the worker (senders only)
+        senders: "list[tuple[int, int, int, WorkerState, Iterator[TaskState]]]" = []
+        recipients: "list[tuple[int, int, int, WorkerState]]" = []
+
+        # Output: [(sender, recipient, task), ...]
+        msgs: "list[tuple[WorkerState, WorkerState, TaskState]]" = []
+
+        # optimistic memory = RSS - unmanaged memory that appeared over the last 30
+        # seconds (distributed.worker.memory.recent_to_old_time).
+        # This lets us ignore temporary spikes caused by task heap usage.
+        memory_by_worker = [
+            (ws, getattr(ws.memory, MEMORY_REBALANCE_MEASURE)) for ws in workers
+        ]
+        mean_memory = sum(m for _, m in memory_by_worker) // len(memory_by_worker)
 
-                worker_bytes = {
-                    ws: sum(ts.get_nbytes() for ts in v)
-                    for ws, v in tasks_by_worker.items()
-                }
+        for ws, ws_memory in memory_by_worker:
+            if ws.memory_limit:
+                half_gap = int(MEMORY_REBALANCE_HALF_GAP * ws.memory_limit)
+                sender_min = MEMORY_REBALANCE_SENDER_MIN * ws.memory_limit
+                recipient_max = MEMORY_REBALANCE_RECIPIENT_MAX * ws.memory_limit
+            else:
+                half_gap = 0
+                sender_min = 0.0
+                recipient_max = math.inf
+
+            if (
+                ws._has_what
+                and ws_memory >= mean_memory + half_gap
+                and ws_memory >= sender_min
+            ):
+                # This may send the worker below sender_min (by design)
+                nbytes = mean_memory - ws_memory  # negative
+                senders.append(
+                    (nbytes, nbytes + half_gap, id(ws), ws, iter(ws._has_what))
```

Comment:
it's already documented above (see definition of senders) but I will add a reminder here.
You need to store an iter object because you want to go through all tasks of each sender from oldest to newest, but you want to pick tasks from senders that need to shed off more first in case there's not enough capacity in the receivers.

We're not abusing anything - the lifetime of the iter is contained within a non-async function in a single-threaded process.

Reply:
With abusing I rather mean: Are relying on this iterator to be initialized here for the algorithm to finish or is the code section below self contained and will terminate regardlessly

Reply:
Does the code rely on this iterator to be exhausted? IIUC the iterator is used in multiple loop iterations and the termination condition depends on this iterator to eventually be empty.

Put differently: Would the below algorithm work if this was a `list(ws._has_what)` instead of an `iter`?

Reply:
> Does the code rely on this iterator to be exhausted?

Iterators will typically *not* be exhausted. The algorithm continues iterating on the keys of a worker until

a) all keys are exhausted (by having been moved to other workers, or by having been skipped because no suitable receivers were found), OR
b) the worker has dropped below the threshold.

Use case (a) exclusively happens when the unmanaged memory + keys that cannot be moved on their own are sufficient to go above the threshold.

> Would the below algorithm work if this was a list(ws._has_what) instead of an iter?

if you mean creating a list of the keys and then popping one element from the left of the list until empty, yes it would have been the same just marginally slower and less readable.